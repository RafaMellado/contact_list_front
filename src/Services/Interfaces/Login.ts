export interface LoginRequestBody {
  email: string;
  password: string;
}

export interface LoginResponse {
  token: string;
  username: string;
  exp: string;
}

export interface LoginError {
  error: string;
}
