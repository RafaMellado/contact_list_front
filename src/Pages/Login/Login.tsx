import { FormEvent, useRef } from "react";
import { Button, Card, Form } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import AuthenticationService from "../../Services/AuthenticationService";

export function Login() {
  const { t } = useTranslation();

  const TRANSLATIONS: string = "login";

  const email = useRef<HTMLInputElement>(null);
  const password = useRef<HTMLInputElement>(null);

  const login = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    await AuthenticationService.login({
      email: String(email?.current?.value),
      password: String(password?.current?.value),
    });
  };

  return (
    <div className="d-flex justify-content-center">
      <Card className="login">
        <Card.Body>
          <Form
            onSubmit={login}
            className="d-flex flex-column align-items-center"
          >
            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.email`)}</Form.Label>
              <Form.Control
                as="input"
                type="email"
                placeholder={t<string>(`${TRANSLATIONS}.email`)}
                ref={email}
              />
            </Form.Group>

            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.password`)}</Form.Label>
              <Form.Control
                as="input"
                type="password"
                placeholder={t<string>(`${TRANSLATIONS}.password`)}
                ref={password}
              />
            </Form.Group>

            <Button variant="primary" type="submit">
              {t<string>(`${TRANSLATIONS}.submit`)}
            </Button>
          </Form>
        </Card.Body>
      </Card>
    </div>
  );
}
