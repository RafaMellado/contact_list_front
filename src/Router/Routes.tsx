import { Routes as Switch, Route } from "react-router-dom";
import { Login } from "../Pages/Login/Login";
import { Home } from "../Pages/Home/Home";
import { ContactBookNew } from "../Pages/ContactBooks/New";
import { ContactNew } from "../Pages/Contacts/New";
import { ContactBookShow } from "../Pages/ContactBooks/Show";
import { ContactShow } from "../Pages/Contacts/Show";
import { ContactEdit } from "../Pages/Contacts/Edit";
import { PrivateRoute } from "./PrivateRoute";
import { PublicRoute } from "./PublicRoute";

export function Routes() {
  return (
    <Switch>
      <Route path="/" element={<PublicRoute RouteComponent={Login} />} />
      <Route path="/home" element={<PrivateRoute RouteComponent={Home} />} />

      <Route
        path="/contact-book/new"
        element={<PrivateRoute RouteComponent={ContactBookNew} />}
      />
      <Route
        path="/contact-book/show/:id"
        element={<PrivateRoute RouteComponent={ContactBookShow} />}
      />

      <Route
        path="/contact-book/:contactBookId/contact/new"
        element={<PrivateRoute RouteComponent={ContactNew} />}
      />

      <Route
        path="/contact/:id"
        element={<PrivateRoute RouteComponent={ContactShow} />}
      />

      <Route
        path="/contact/:id/edit"
        element={<PrivateRoute RouteComponent={ContactEdit} />}
      />
    </Switch>
  );
}
