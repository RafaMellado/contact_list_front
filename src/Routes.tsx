import React from "react";
import { Routes as Switch, Route } from "react-router-dom";
import { Login } from "./Pages/Login/Login";

export function Routes() {
  return (
    <Switch>
      <Route path="/" element={<Login />}></Route>
    </Switch>
  );
}
