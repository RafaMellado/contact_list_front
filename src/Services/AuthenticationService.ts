import Cookies from "universal-cookie";
import { LoginRequestBody, LoginResponse } from "./Interfaces/Login";
import LoginRepository from "./Repository/LoginRepository";

class AuthenticationService {
  async login(data: LoginRequestBody) {
    const response = await LoginRepository.login(data);

    if (response.token) {
      this.setAuthCookie(response);

      window.location.href = "/home";
    }
  }

  setAuthCookie(data: LoginResponse) {
    const cookies = new Cookies();

    cookies.set(
      "user",
      { token: data.token, username: data.username },
      { expires: new Date(data.exp) }
    );
  }
}

export default new AuthenticationService();
