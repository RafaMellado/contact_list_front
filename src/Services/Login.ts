import { LoginData, LoginResponse } from "./Interfaces/Login";
import RestService from "./RestService";
import Cookies from "universal-cookie";

class LoginService {
  async login(loginData: LoginData) {
    const response = await RestService.post("/auth/login", loginData);
    const data: LoginResponse = await response.json();

    if (data.token) {
      this.setCookie(data);
    }

    return !!data.token;
  }

  setCookie(data: LoginResponse) {
    const cookies = new Cookies();

    cookies.set(
      "user",
      { token: data.token, username: data.username },
      { expires: new Date(data.exp) }
    );
  }
}

export default new LoginService();
