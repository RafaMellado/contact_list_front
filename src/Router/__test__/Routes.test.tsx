import { Routes } from "../Routes";
import { act, render } from "@testing-library/react";
import { Login } from "../../Pages/Login/Login";
import { Router } from "react-router-dom";
import AuthenticationService from "../../Services/AuthenticationService";
import { createMemoryHistory } from "history";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("../../Services/AuthenticationService", () => ({
  isLogged: jest.fn(),
}));

describe(`<${Routes.name} />`, () => {
  const history = createMemoryHistory();

  const factoryComponent = () =>
    render(
      <Router location="/" navigator={history}>
        <Routes />
      </Router>
    );

  test("renders correctly the component", () => {
    const component = factoryComponent();

    act(() => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
