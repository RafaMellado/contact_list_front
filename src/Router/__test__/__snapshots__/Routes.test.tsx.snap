// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Routes /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="login card"
    >
      <div
        class="card-body"
      >
        <form
          class="d-flex flex-column align-items-center"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              login.email
            </label>
            <input
              class="form-control"
              data-testid="login-email"
              placeholder="login.email"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              login.password
            </label>
            <input
              class="form-control"
              data-testid="login-password"
              placeholder="login.password"
              type="password"
            />
          </div>
          <button
            class="btn btn-primary"
            data-testid="login-submit"
            type="submit"
          >
            login.submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
